### ISSUE TITLE
[backend] feat: endpoint to extract one contributor's cyclic inconsistencies

### ISSUE THREAD

**Issue by lfaucon** [tournesol-app/tournesol#671]
It may happen that a contributor compares contents A, B and C on criteria X and submits that A > B, B > C and C > A. This is an impossibility. We should open an endpoint that helps contributors extract such inconsistencies in their comparisons to make it possible for them to correct their ratings.

The endpoint may be:

`/users/me/inconsistencies/{pollname}`

which returns a list of inconsistencies. Each inconsistencies should contain the name of the incorrectly Criteria, the three ComparisonCriteriaScores, the three Entities and their metadata.

**Comment by glerzing** [comment]
I don't agree with the needs.

Contributors probably won't enjoy having to fix inconsistencies. This could feel a bit like a failure, and make them more careful with new comparisons, whereas we want it to be pleasant and effortless. I think this goes against user experience.

And inconsistencies are normal to human reasoning. It can probably even be a valuable information for machine learning, to be able to estimate how reliable are a contributor's ratings. And maybe there is a value in keeping the database "natural", and let any enterprise importing the database make their own pre-processing. But I am not a machine learning expert.

What are others' opinions on this?

**Comment by Omicron666** [comment]
On the detection of inconsistencies themselves, you actually need to check for cycles. Not just 3 elements or you will miss many of them.
Also thinking about cycles is probably easier.

**Comment by lfaucon** [comment]
After a groupd discussion, we agreed that it is important for a contributor to have access to his or her list of inconsistencies in order to correct them because having inconsistencies reduces the contributor's impact on the aggregated scores

**Comment by lfaucon** [comment]
A second interesting type of inconsistencies is the differences between ContributorRatingCriteriaScores and the Comparison

@lenhoanglnh Can you provide the mathy formula to extract this type of inconsistencies?

**Comment by glerzing** [comment]
I am quite lost with this feature.

For the cyclic inconsistencies, the easy, naive approach's complexity for 3-elements-cycles is in O(n³) => not great
The paragraph 3.2 of the paper https://www.researchgate.net/publication/338671588_Inconsistency_indices_for_incomplete_pairwise_comparisons_matrices
suggests a complicated formula for calculating cyclic inconsistencies in incomplete PC. It looks like its time complexity is in O(n²), and it should also work for cyclic inconsistencies of length > 3. Is that what you want me to implement?

For the 2nd proposed measure of inconsistencies, linking the Comparison of 2 entities and the subtraction of the ContributorRatingCriteriaScores of both entities, I didn't find a meaningful formula outputting the inconsistency.

**Comment by glerzing** [comment]
While I agree that not all cyclic inconsistencies have a length of 3, I didn't find an efficient algorithm to find longer cyclic inconsistencies. It would make a lot of cycles to check, and besides it would be more complicated to process and display the results in the frontend.
There isn't a lot of literature on pairwise comparison matrices that are incomplete (and it already looks surprisingly complicated even for complete matrices).
So I will probably implement only the length 3 inconsistency cycles search. 
(I still believe though there is a risk that solving inconsistencies would feel tedious for contributors, especially if it leads to new inconsistencies)

For the 2nd proposed measure of inconsistencies, linking the Comparison of 2 entities and the subtraction of the ContributorRatingCriteriaScores of both entities, I didn't find a meaningful formula outputting the inconsistency. Do you think there is such a formula, @lenhoanglnh ?

**Comment by lfaucon** [comment]
Our model is (maybe "used to be") an adaptiation of https://en.wikipedia.org/wiki/Bradley%E2%80%93Terry_model 

I believe we should have roughly:

```
ComparisonScore = -10 + 20 * 1 / ( 1 + exp( lambda * ( ratingLeft - ratingRight ) ) )
```

This leads for example to `ComparisonScore = -10` when `ratingLeft >> ratingRight` and `ComparisonScore = 10` when `ratingLeft << ratingRight`.

We have to figure out the `lambda` that matches the current algorithm before we can use this formula well.

**Comment by lenhoanglnh** [comment]
We have new algorithms, with a transformation of comparisons (in -10 to 10) into score differences (unbounded in principle).
In the following R_max = 10.
<img width="1105" alt="Screen Shot 2022-03-26 at 21 59 49" src="https://user-images.githubusercontent.com/10453308/160256893-1cb641f3-e5ae-4c94-ace4-cf3f415f065f.png">
Using Louis' notation, this leads us to define:

```
RatingDifference = RatingLeft - RatingRight
RepresentativeComparisonScore = - RatingDifference / sqrt( 1+ RatingDifference^2 )
Inconsistency = abs( RepresentativeComparisonScore - SubmittedComparisonScore )
```
In particular, an inconsistency will be large if `Inconsistency >= 1`.

NB: This inconsistency should eventually be modified to account for the uncertainty on `RatingLeft` and `RatingRight`, which will be available once Adrien updates the algorithms. The basic idea is to either add or subtract the uncertainties `RatingUncertaintyLeft + RatingUncertaintyRight` to `RatingDifference`, so that `Inconsistency` is minimized.

**Comment by Omicron666** [comment]
For cycle detection

@glerzing There are cyclic detection algorithms for directed graphs (exact weights of comparisons don't matter) like [Depth First Search](https://www.geeksforgeeks.org/detect-cycle-in-a-graph/), with linear complexity

Your matrix here is currently the support to describe your graph:
- sign of comparison value give edge direction
- zero value mean two nodes A,B are considered equals (for that given attribute) (this is maybe the non-usual case. Imho, in that case, consider two directed edges for A<->B, and drop trivial 2-cycles)

If your matrix is antisymmetric, you can skip all strictly negative values on-the-fly.

**Comment by glerzing** [comment]
Thanks for the feedback

**Comment by glerzing** [comment]
This is the kind of situations that makes me fear a denial of service if we implement the search of cycles of length > 3. With relatively few entities and comparisons, we can arrive to a situation 

![image](https://user-images.githubusercontent.com/47578089/160474755-937132ad-7487-48f2-907a-add968bc342a.png)

**Comment by glerzing** [comment]
For the cyclic inconsistencies, I suggest to make 2 API endpoints: one that counts the number of length 3 cycles and lists some of them, and one that returns any cycle of any length.

The rationale is that the number of arbitrary-length cycles can get exponential, and counting them looks like it is NP-complex (https://www.quora.com/How-do-I-count-the-number-of-cycles-in-a-directed-graph).
[cycles_exponentiality.odg](https://github.com/tournesol-app/tournesol/files/8506068/cycles_exponentiality.odg)


Counting the length-3 cycles may have a complexity of O(nb_criteria * nb_entities * nb_comparisons_per_entity ² ), so I'm not sure it will be fast enough. But I think it is important to let the possibility to have a count, and not just blindly return any cycle.

I prefer to ignore the zero-value comparisons than to consider it a bidirectional edge. It is simpler, and a zero-value comparison could be considered to be due to the fact that the comparison scores are on integers and not on floats.

**Comment by lfaucon** [comment]
I'm happy with your proposition of having one endpoint for 3-cycles and one for arbitrary length. Let's do it :)

The complexity of `O(nb_criteria * nb_entities * nb_comparisons_per_entity ² )` seems to be acceptable. If it is slow for some super-users, then we could add caching, or precomputing, or giving the user the option to request the analysis, then retrieve the results of the inconsistencies later. Let's implement the simplest version first and look into the optimization if it is needed later.

> I prefer to ignore the zero-value comparisons than to consider it a bidirectional edge. It is simpler, and a zero-value comparison could be considered to be due to the fact that the comparison scores are on integers and not on floats.

I don't really understand what you mean by "zero-value comparison". The connection you have to consider for the algorithm are in any case the ComparisonCriteriaScore

**Comment by Omicron666** [comment]
You don't need to enumerate every possible route combinations (that may indeed grow exponentially especially if intentional!) but a subgraph including involved nodes and their respective edges.
Frontend can always decide to enumerate all or some of problematic cycles.

Eventually, we could forbid users from submitting more comparaisons after some hard limit (after a warning soft limit), to prevent him to get him/her in a situation s/he won't be cognitively able to correct his/her comparaisons, that may not be used by tournesol anyway in that case !?,

**Comment by Omicron666** [comment]
Also you may not want to allow users to have more than a couple of inconsistencies, or, correcting one may just create new other ones in cascades.

**Comment by Omicron666** [comment]
Could other alternative solution be to weight comparisons less overtime, to help auto-correct inconsistencies?...

**Comment by lfaucon** [comment]
> Also you may not want to allow users to have more than a couple of inconsistencies, or, correcting one may just create new other ones in cascades.

Our main goal is to collect a large database, so we don't want to block users to submit new comparison in any situatino eve if they are inconsistent. It's actually really interesting for research to evaluate how insconsistent users are and also to build system that still make publicly beneficial decisions based on inconsistent judgements from the public!

**Comment by glerzing** [comment]
By "zero-value comparison" I meant comparisons with a score of 0 (no entity clearly preferred).

**Comment by lfaucon** [comment]
> By "zero-value comparison" I meant comparisons with a score of 0 (no entity clearly preferred).

Of course, I should have understood! So I confirm these should be ignored and not consider any directed edge from them.

**Comment by Omicron666** [comment]
> > 
> 
> Of course, I should have understood! So I confirm these should be ignored and not consider any directed edge from them.

That will be simpler for cycle computation.
But there is clearly semantic distinction between "two videos are equivalent" (i) and "two videos are not compared" (ii)

If you have 
A > B > C > D = A
you will miss a cycle. (... and how inconsistent one might be, and you don't want to miss that! ;-) )

Other than that, you might reconsider UI, to remove checkbox if you consider (i) and (ii) are equivalent, to avoid making people check/uncheck multiple boxes for people unaware. People will get aware of their uselessness in the end, so they'll start ignoring them (and you won't be able to rely on distinction of (i)/(ii) anymore in the end).

Here a quick proposal for that case, but that may be object of dedicated ticket.

![image](https://user-images.githubusercontent.com/23405122/163865352-588f67fd-1c07-4a30-b344-3d2f32e4c2ec.png)

↓

![image](https://user-images.githubusercontent.com/23405122/163865134-82e08b5f-62ba-48bb-95c3-6b6f6dc12234.png)

**Comment by glerzing** [comment]
I don't see an issue with having no significant preference between the 2 entities and keeping the value 0. I don't think banning the value 0 is a good thing.
But I agree that it's probably not rare to forget to set a cursor, and to let it to 0. Maybe another possibility would be to have the checkboxs off by default, and when the contributor clicks anywhere on a criterion grid (even on the middle), the corresponding checkbox activates and the criterion is counted. This way it's easier to avoid forgetting any criterion (and to compare on only some of the criteria).

**Comment by Omicron666** [comment]
Yes, in both case of whatever is chosen, UI/UX would benefit iterations.

 ```mermaid
graph TD
    A -->|0.4| B -->|0.4| C
    A -->|0.8| C
```
(just trying new graph feature... ^^)

- I don't understand at what step those (normalized?) values get rounded.
- Can't we compute cycle inconsistencies directly on user input values? That may be the only numbers we expose him to, in the end.

**Comment by glerzing** [comment]
> * I don't understand at what step those (normalized?) values get rounded.
I meant, the preferences of 0.4 and 0.8 would be your mental preferences, but when submitting a comparison you have to chose between integers (tiers on the comparisons slider) from -10 to 10. So, you submit 0 or 1. You would get A = B = C < A

**Comment by Omicron666** [comment]
> > *
> 
> I meant, the preferences of 0.4 and 0.8 would be your mental preferences, but when submitting a comparison you have to chose between integers (tiers on the comparisons slider) from -10 to 10. So, you submit 0 or 1. You would get A = B = C < A

Maybe that would make it fall in trivial cycles (like 2-cycles) that need to be ignored. (Count non-zero weighted edges only, to compute a current cycle length to decide to ignore such a cycle)

**Comment by lfaucon** [comment]
Thanks for the discussion ! 

Indeed there are many ways for comparisons to be inconsistent. 

However I prefer to keep the cycle detection as simple as possible and to only show the most obvious inconsitencies of the type A > B > C > A .

Other inconsistencies will be found by the other type of incosistencies that @glerzing is working on the differences between the ContributorRatings and the Comparisons.

To be agile here, I would find best to have this simpler version implemented first, then collect feedback from users and iterate the next improvements based on the feedback.

**Comment by glerzing** [comment]
Hello, I hesitate between these 2 API outputs for the length-3 cycles, 1 where the results are sorted by criterion (I called a "comparisons trio" any 3 entities that have a comparison to each other) :

  "per_criterion": [
    {
      "criterion": "string",
      "cycles_count": 0,
      "comparison_trios_count": 0,
      "cycles": [
        {
          "entity_1_uid": "string",
          "entity_2_uid": "string",
          "entity_3_uid": "string"
        }
Hello, I hesitate between these 2 API outputs for the length-3 cycles, 1 where the results are sorted by criterion (I called a "comparisons trio" any 3 entities that have a comparison to each other) :

  "per_criterion": [
    {
      "criterion": "string",
      "cycles_count": 0,
      "comparison_trios_count": 0,
      "cycles": [
        {
          "entity_1_uid": "string",
          "entity_2_uid": "string",
          "entity_3_uid": "string"
        }
      ]
    }
  ]
  
  
and one where the results are not sorted per criterion, but there is a criterion API parameter that allows to filter according to one criterion:

  "cycles_count": 0,
  "comparison_trios_count": 0,
  "cycles": [
    {
      "criterion": "string",
      "entity_1_uid": "string",
      "entity_2_uid": "string",
      "entity_3_uid": "string"
    }
  ]
 
The advantage of the first one is that, if you are interested by the results per criterion for all criteria, you can call it only once whereas with the 2nd one you would need to call it as many times as there are criteria. The 2nd one is simpler to process by the frontend, and may be better if you don't need the results per criterion or you don't care calling 10 times the API.




Similarly for the score inconsistencies, that I plan to change, I hesitate between :

  "per_criterion": [
    {
      "criterion": "string",
      "mean_inconsistency": 0,
      "nb_inconsistent_comparisons": 0,
      "nb_comparisons": 0,
      "results": [
		{
		  "inconsistency": 0,
		  "entity_1_uid": "string",
		  "entity_2_uid": "string",
		  "comparison_score": 0,
		  "entity_1_rating": 0,
		  "entity_2_rating": 0
		}
	  ]
    }
  ]


And, with the API input criterion filter parameter :

  "mean_inconsistency": 0,
  "nb_inconsistent_comparisons": 0,
  "nb_comparisons": 0,
  "results": [
    {
      "criterion": "string",
      "inconsistency": 0,
      "entity_1_uid": "string",
      "entity_2_uid": "string",
      "comparison_score": 0,
      "entity_1_rating": 0,
      "entity_2_rating": 0
    }
  ]

In this case I think the 2nd option is better. But again, if you want to be able to compare your results on the different criteria (which I think is a nice feature), you would have to call the API once for each criterion. 

If you think there is a better solution, or if you want to add or remove a parameter, don't hesitate to suggest.

**Comment by lfaucon** [comment]
I prefer the ones with a single list in both cases even if it comes with slightly more redundancy. The simpler structure will be easier to manipulate on the frontend and also to generate in the backend.

>  there is a criterion API parameter that allows to filter according to one criterion

This is nice to have, but I don't think we will have a lot of usecases for it.

> The advantage of the first one is that, if you are interested by the results per criterion for all criteria, you can call it only once whereas with the 2nd one you would need to call it as many times as there are criteria. The 2nd one is simpler to process by the frontend, and may be better if you don't need the results per criterion or you don't care calling 10 times the API.

I don't really understand this. Can the return be a list including all criterias? For example:

```
{
  "count": 0,
  "results": [
    {
      "criterion": "largely_recommended",
      "entity_1_uid": "videoA",
      "entity_2_uid": "videoB",
      "entity_3_uid": "videoC"
    },
    {
      "criterion": "better_habits",
      "entity_1_uid": "videoA",
      "entity_2_uid": "videoE",
      "entity_3_uid": "videoB"
    }
    {
      "criterion": "important",
      "entity_1_uid": "videoA",
      "entity_2_uid": "videoF",
      "entity_3_uid": "videoG"
    }
  ]
}
```

I would also maybe recommend to use `"count"` and `"results"` like for other list response. It makes more sense later to do `fetchCycles().results` thant `fetchCycles().cycles`

**Comment by glerzing** [comment]
if you have a few minutes, maybe we can settle this in the chat

**Comment by glerzing** [comment]
After discussion, the chosen output for cyclic inconsistencies is :

```
  "count": "integer",
  "results": [
        {
          "criterion": "string",
          "entity_1_uid": "string",
          "entity_2_uid": "string",
          "entity_3_uid": "string"
        }
      ]
  "stats": [
        {
          "criterion": "string",
          "cycles_count": "integer",
          "comparison_trios_count": "integer",
        }
      ]
```

and for score inconsistencies:

```
  "count": "integer",
  "results": [
        {
            "criterion": "string",
            "inconsistency": "float",
            "entity_1_uid": "string",
            "entity_2_uid": "string",
            "comparison_score": "float",
            "entity_1_rating": "float",
            "entity_2_rating": "float"
        }
      ]
  "stats": [
        {
            "criterion": "string",
            "mean_inconsistency": "float",
            "nb_inconsistent_comparisons": "integer",
            "nb_comparisons": "integer",
        }
      ]
```
For results, there will be as many elements as inconsistencies, and for stats, there will be as many elements as criteria.
